File: src/state.js
// Project state: everything the tool can save, load and export.

import { createMaterials, deserializeMaterials, serializeMaterials, invalidateSamplerCache } from './sampler.js';
import { defaultShading } from './shading.js';
import { defaultClassLimits, defaultSpeciesList, makeSpecies } from './species.js';
import { defaultWorldConfig } from './world.js';
import { defaultCivConfig, mergeCivConfig } from './civ/config.js';
import { deepClone } from './util.js';

export const STORAGE_KEY = 'grow.project.v1';

// Bumped when the world model changes shape. Version 2 turned the world from a
// side view strip into a 2.5D area, so a version 1 world config is discarded.
// Version 3 added the settlement: its own map, people, economy and technology.
export const STATE_VERSION = 3;

export function createState() {
  return {
    version: STATE_VERSION,
    seed: 20260815,
    materials: createMaterials(),
    shading: defaultShading(),
    species: defaultSpeciesList(),
    classLimits: defaultClassLimits(),
    world: defaultWorldConfig(),
    sim: { speed: 1, running: true, rasterBudget: 12, tickHz: 20 },
    civ: defaultCivConfig(),
  };
}

export function serializeState(state) {
  return {
    version: state.version,
    seed: state.seed,
    materials: serializeMaterials(state.materials),
    shading: deepClone(state.shading),
    species: deepClone(state.species),
    classLimits: deepClone(state.classLimits),
    world: deepClone(state.world),
    sim: deepClone(state.sim),
    civ: deepClone(state.civ),
  };
}

export function deserializeState(data) {
  const fresh = createState();
  const stale = (data.version || 1) < STATE_VERSION;
  const state = {
    version: STATE_VERSION,
    seed: data.seed ?? fresh.seed,
    materials: data.materials ? deserializeMaterials(data.materials) : fresh.materials,
    shading: { ...fresh.shading, ...(data.shading || {}) },
    // Older projects may miss newer fields; makeSpecies fills the gaps.
    species: (data.species || fresh.species).map((sp) => makeSpecies(sp)),
    classLimits: { ...fresh.classLimits, ...(data.classLimits || {}) },
    world: stale ? fresh.world : { ...fresh.world, ...(data.world || {}) },
    sim: { ...fresh.sim, ...(data.sim || {}) },
    civ: mergeCivConfig(data.civ),
  };
  invalidateSamplerCache();
  return state;
}

export function saveLocal(state) {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(serializeState(state)));
    return true;
  } catch (err) {
    console.warn('save failed', err);
    return false;
  }
}

export function loadLocal() {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return null;
    return deserializeState(JSON.parse(raw));
  } catch (err) {
    console.warn('load failed', err);
    return null;
  }
}

export function clearLocal() {
  try {
    localStorage.removeItem(STORAGE_KEY);
  } catch (err) {
    console.warn('clear failed', err);
  }
}

File: rust/src/state.rs
//! Project state: everything the tool can save, load and export.
//!
//! The JSON shape is the one the project has always used, so a file exported
//! before this rewrite still loads. Every section carries its own defaults, and
//! serde fills a missing field from them, which is what lets an older project
//! pick up parameters that did not exist when it was saved.

use serde::{Deserialize, Serialize};

use crate::art::ArtLibrary;
use crate::civ::config::{CivConfig, SimSettings};
use crate::sampler::Materials;
use crate::shading::Shading;
use crate::species::{default_species_list, ClassLimits, Species};
use crate::world::WorldConfig;

pub const STORAGE_KEY: &str = "grow.project.v1";

/// Bumped when the world model changes shape. Version 2 turned the world from a
/// side view strip into a 2.5D area, so a version 1 world config is discarded.
/// Version 3 added the settlement: its own map, people, economy and technology.
pub const STATE_VERSION: u32 = 3;

fn default_version() -> u32 {
    STATE_VERSION
}

fn default_seed() -> u32 {
    20260815
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct State {
    pub version: u32,
    pub seed: u32,
    pub materials: Materials,
    /// Sprite sheets drawn in the tool, which settler motions can be pointed
    /// at instead of at a dropped image.
    pub art: ArtLibrary,
    pub shading: Shading,
    pub species: Vec<Species>,
    pub class_limits: ClassLimits,
    pub world: WorldConfig,
    pub sim: SimSettings,
    pub civ: CivConfig,
}

impl Default for State {
    fn default() -> Self {
        State {
            version: default_version(),
            seed: default_seed(),
            materials: Materials::new(),
            art: ArtLibrary::default(),
            shading: Shading::default(),
            species: default_species_list(),
            class_limits: ClassLimits::default(),
            world: WorldConfig::default(),
            sim: SimSettings::default(),
            civ: CivConfig::default(),
        }
    }
}

impl State {
    pub fn new() -> Self {
        State::default()
    }

    pub fn to_json(&self) -> String {
        serde_json::to_string(self).unwrap_or_else(|_| "{}".to_string())
    }

    pub fn to_pretty_json(&self) -> String {
        serde_json::to_string_pretty(self).unwrap_or_else(|_| "{}".to_string())
    }

    /// Loads a project, upgrading it on the way in.
    pub fn from_json(raw: &str) -> Result<State, String> {
        let mut state: State = serde_json::from_str(raw).map_err(|e| e.to_string())?;
        let stale = state.version < STATE_VERSION;
        if stale {
            // The world model changed shape, so an old world config is dropped
            // rather than half applied.
            state.world = WorldConfig::default();
        }
        state.version = STATE_VERSION;
        if state.species.is_empty() {
            state.species = default_species_list();
        }
        state.art.fit();
        state.materials.ensure_role_samplers();
        state.materials.invalidate();
        Ok(state)
    }

    pub fn species_index(&self, id: &str) -> Option<usize> {
        self.species.iter().position(|s| s.id == id)
    }

    pub fn find_species(&self, id: &str) -> Option<&Species> {
        self.species.iter().find(|s| s.id == id)
    }
}
